core/validation/validate.py

```diff
@@ -169,6 +170,38 @@ def validate_uniques(
     return non_unique_failures
 
 
+def validate_unique_composite_keys(
+    workbook: dict[str, pd.DataFrame], sheet_name: str, composite_keys: tuple
+) -> list[vf.NonUniqueCompositeKeyFailure]:
+    """
+    Validates the uniqueness of composite keys across multiple columns.
+
+    :param workbook: A dictionary containing sheet names as keys and corresponding pandas DataFrames as values.
+    :param sheet_name: The name of the worksheet to be validated.
+    :param composite_keys: A list of tuples, where each tuple contains the column names
+                           that should have combined uniqueness.
+    :return: A list of non-unique composite key failures, if any exist.
+    """
+
+    sheet = workbook[sheet_name]
+    non_unique_composite_key_failures = []
+
+    # removing duplicates as to avoid index mismatch
+    sheet.drop_duplicates(subset=list(composite_keys), inplace=True)
+    # filter dataframe by these columns and find duplicated rows including NaN/Empty
+    mask = sheet[list(composite_keys)].duplicated(keep=False)
+    mask.drop_duplicates(inplace=True)
```

Comment:
removing the duplicates before identifying the duplicates breaks the intended functionality because no duplicates will ever be detected.

the test for this function `test_validate_unique_composite_keys_invalid` needs correcting too. At the moment the test iterates over each failure returned, making assertions on each failure - but this means if no failures are returned then we just never make an assertion and test passes incorrectly. Need to be really careful with things like this. I would opt to be verbose and explicit in tests and just write out each assertion without using loops etc. It will help with readability and also protect against problems like this.